src/IonParserBinaryRaw.ts

```diff
@@ -213,28 +217,40 @@ export class ParserBinaryRaw {
         return LongInt.fromIntBytes(Array.prototype.slice.call(bytes), sign);
     }
 
-    private readUnsignedInt() : number {
-        var v = 0, b;
+    static readUnsignedIntFrom(input: BinarySpan, numberOfBytes: number) {
+        let value = 0, bytesRead = 0, byte;
+        if (numberOfBytes < 1)
+            return 0;
 
-        // they have to tell us the length
-        if (this._len < 1) return 0;
+        while (bytesRead < numberOfBytes) {
+            byte = input.next();
+            bytesRead++;
+            value = value << 8;
+            value = value | (byte & 0xff);
+        }
 
-        // shift in all but the last byte (we've already read the first)
-        while (this._len > 0) {
-            b = this._in.next();
-            this._len--;
-            v = v << 8;
-            v = v | (b & 0xff);
+        // We overflowed
+        if (value < 0) {
```

Comment:
is there a chance to overflow twice so it's back to positive?

Reply:
In extreme cases, yes. I've modified the check to also fail if the number of bytes read was >4, accounting for all overflow cases beyond 32 bits.